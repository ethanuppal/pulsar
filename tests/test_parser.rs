#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use pulsar_frontend::{lexer::Lexer, parser::Parser};
    use pulsar_utils::{error::ErrorManager, loc::Source};
    use std::{cell::RefCell, fs, rc::Rc};

    fn read(filename: &str) -> Rc<Source> {
        Source::file(
            filename.into(),
            fs::read_to_string(filename)
                .expect(format!("Could not read file: {}", filename).as_str())
        )
    }

    fn parser_output(
        filename: &str, error_manager: RRC<ErrorManager>
    ) -> String {
        let source = read(filename);
        let lexer = Lexer::new(source, error_manager.clone());
        let tokens: Vec<_> = lexer.into_iter().collect();
        let parser = Parser::new(tokens, error_manager.clone());
        let mut output = String::new();
        for node in parser {
            output.push_str(&format!("{}\n", node));
        }

        let mut buffer = Vec::new();
        if error_manager.borrow().has_errors() {
            error_manager
                .borrow_mut()
                .consume_and_write(&mut buffer)
                .unwrap_or_default();
        }
        output.push_str(String::from_utf8(buffer).unwrap().as_str());
        output
    }

    use paste::paste;

    macro_rules! generate_test {
        ($num:expr) => {
            paste! {
                #[test]
                fn [<test_parser_ $num>]() {
                    let error_manager = ErrorManager::with_max_count(10);
                    assert_snapshot!(parser_output(
                        &format!("tests/data/parser{}.plsr", $num),
                        error_manager
                    ));
                }
            }
        };
    }

    generate_test!(1);
    generate_test!(2);
    generate_test!(3);
    generate_test!(4);
    generate_test!(5);
    generate_test!(6);
    generate_test!(7);
    generate_test!(8);
    generate_test!(9);
    generate_test!(10);
}
