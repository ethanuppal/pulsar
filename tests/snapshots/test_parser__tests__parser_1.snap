---
source: tests/test_parser.rs
expression: "parser_output(&format!(\"tests/data/parser{}.plsr\", 1), error_manager)"
---
func main() -> Unit {
    return
}
error[E0006]: tests/data/parser1.plsr:6:5: Invalid token at the start of a statement
     │  
   5 │  func foo() {
   6 │      fun bad_nested() {
   7 │  
     │  

info[I0003]: tests/data/parser1.plsr:6:9: Attempting to recover understanding of code
     │  
   5 │  func foo() {
   6 │      fun bad_nested() {
     │          └───────── Seeking end of function body
   7 │  
     │  

info[I0003]: tests/data/parser1.plsr:8:6: Attempting to recover understanding of code
     │  
   7 │  
   8 │      } 
     │       └ Seeking top-level construct
   9 │  }
     │  

error[E0005]: tests/data/parser1.plsr:12:5: Unexpected top-level construct
     │  
  11 │  func foo() {
  12 │      func nested() {
  13 │          
     │  
Suggestion: Did you mean to place it at the top level?

info[I0003]: tests/data/parser1.plsr:12:10: Attempting to recover understanding of code
     │  
  11 │  func foo() {
  12 │      func nested() {
     │           └───── Seeking end of function body
  13 │          
     │  

info[I0003]: tests/data/parser1.plsr:14:6: Attempting to recover understanding of code
     │  
  13 │          
  14 │      } 
     │       └ Seeking top-level construct
  15 │  }
     │  

error[E0003]: tests/data/parser1.plsr:19:1: Expected 'identifier' for parameter name in `missing_close`
     │  
  18 │  
  19 │  func missing_close_2() {
     │  └─── Received 'func' here
     │  

info[I0003]: tests/data/parser1.plsr:19:1: Attempting to recover understanding of code
     │  
  18 │  
  19 │  func missing_close_2() {
     │  └─── Seeking top-level construct
     │  

error[E0002]: tests/data/parser1.plsr:19:25: 
     │  
  18 │  
  19 │  func missing_close_2() { 
     │                          └ Unexpected EOF at end of function body
     │  
   ...
     │  
  18 │  
  19 │  func missing_close_2() { 
     │                         └ { opened here
     │
